datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -614,7 +613,12 @@ impl ExprIdentifierVisitor<'_> {
 impl TreeNodeVisitor for ExprIdentifierVisitor<'_> {
     type N = Expr;
 
-    fn pre_visit(&mut self, _expr: &Expr) -> Result<VisitRecursion> {
+    fn pre_visit(&mut self, expr: &Expr) -> Result<VisitRecursion> {
+        // related to https://github.com/apache/arrow-datafusion/issues/8814
+        // If the expr contain volatile expression or is a case expression, skip it.
+        if matches!(expr, Expr::Case(..)) || is_volatile_expression(expr)? {
```

Comment:
I think this problem could exist in other function/expression like `COALESCE` | `OR`

It can be tracked as a future ticket.

Reply:
The inner expressions in these expressions may be short-circuited.

Reply:
track in #8874

Reply:
I comment in the #8874, Do  we have any method for this rule to make sure the new function/expr or other cases we can't find will not bring the same bug